### PR TITLE
    MGMT-18287: Compare resourceVersions when updating a resource     

### PR DESCRIPTION
**Pull request by ori-amizur**
Compare resource version when performing device updates.  In case of
conflict perform retry when applicable.

### CODE REVIEW

**Comment by avishayt** [comment]
Please link to [MGMT-18287](https://issues.redhat.com/browse/MGMT-18287) in the commit message and PR

File: internal/store/model/resource.go [hunk 1]
```diff
@@ -32,11 +32,11 @@ type Resource struct {
 
 	Annotations pq.StringArray `gorm:"type:text[]"`
 
-	Generation *int64
-
-	CreatedAt time.Time
-	UpdatedAt time.Time
-	DeletedAt gorm.DeletedAt `gorm:"index"`
+	Generation      *int64
+	ResourceVersion *int64
```
Comment (avishayt) [inline internal/store/model/resource.go:36]: Why a pointer if it always exists?

File: internal/store/device.go [hunk 2]
```diff
@@ -107,9 +110,13 @@ func (s *DeviceStore) Create(ctx context.Context, orgId uuid.UUID, resource *api
 	if resource == nil {
 		return nil, flterrors.ErrResourceIsNil
 	}
-	device := model.NewDeviceFromApiResource(resource)
+	device, err := model.NewDeviceFromApiResource(resource)
+	if err != nil {
+		return nil, err
```
Comment (avishayt) [inline internal/store/device.go:115]: flterrors.ErrIllegalResourceVersionFormat ?

File: internal/store/device.go [hunk 3]
```diff
@@ -107,9 +110,13 @@ func (s *DeviceStore) Create(ctx context.Context, orgId uuid.UUID, resource *api
...
+	}
 	device.OrgID = orgId
 	device.Generation = util.Int64ToPtr(1)
+	device.ResourceVersion = lo.ToPtr[int64](1)
 
 	result := s.db.Create(device)
```
Comment (avishayt) [inline internal/store/device.go:121]: Should you call createDevice()?

File: internal/store/device.go [hunk 4]
```diff
@@ -181,88 +188,105 @@ func (s *DeviceStore) Get(ctx context.Context, orgId uuid.UUID, name string) (*a
...
+	device.ResourceVersion = lo.ToPtr[int64](1)
+	if result := s.db.Create(device); result.Error != nil {
+		err := flterrors.ErrorFromGormError(result.Error)
+		return err == flterrors.ErrDuplicateName, err
+	}
+	return false, nil
+}
+
+func (s *DeviceStore) updateDevice(fromAPI bool, existingRecord, device *model.Device, fieldsToUnset []string) (bool, error) {
+	if device.ResourceVersion != nil && lo.FromPtr(existingRecord.ResourceVersion) != lo.FromPtr(device.ResourceVersion) {
```
Comment (avishayt) [inline internal/store/device.go:213]: Move this check down after the check for changing the spec of an owned device

**Review by avishayt — COMMENTED** [review]
Overall looks great. Did a quick first pass. Need to do a more in depth review.

File: internal/store/model/resource.go (hunk 1 above)
Comment (ori-amizur) [reply internal/store/model/resource.go:36]: When the existing and the new resourceVersion(s) are compared, if it is nil (when it comes from the API) we ignore the version check.
In addition, in general I prefer to detect cases when it was not set for some reason.

File: internal/store/device.go (hunk 2 above)
Comment (ori-amizur) [reply internal/store/device.go:115]: It is created here https://github.com/flightctl/flightctl/blob/2c89591e575ea3e52de6a46878404b0c940e4cd2/internal/store/model/device.go#L77
This way the function may return other errors (in the future)

File: internal/store/device.go (hunk 4 above)
Comment (ori-amizur) [reply internal/store/device.go:213]: OK

File: internal/store/device.go [hunk 5]
```diff
@@ -181,88 +185,105 @@ func (s *DeviceStore) Get(ctx context.Context, orgId uuid.UUID, name string) (*a
 	return &apiDevice, nil
 }
 
-func (s *DeviceStore) CreateOrUpdate(ctx context.Context, orgId uuid.UUID, resource *api.Device, fieldsToUnset []string, fromAPI bool, callback DeviceStoreCallback) (*api.Device, bool, error) {
+func (s *DeviceStore) getExistingRecord(name string, orgId uuid.UUID) (*model.Device, bool, error) {
```
Comment (avishayt) [inline internal/store/device.go:188]: Nice helper function.  Let's replace the logic in Get() with this?

**Review by avishayt — COMMENTED** [review]
Looks great, just one comment.  Thank you!

File: internal/store/device.go (hunk 5 above)
Comment (ori-amizur) [reply internal/store/device.go:188]: The semantics is a bit different.  The `getExistingRecord` returns nil error in case the record does not exist.

**Review by avishayt — APPROVED** [review]

**Pull request merged**